src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1296,12 +1342,13 @@ public SolutionCompilationState WithoutFrozenSourceGeneratedDocuments()
                 foreach (var projectId in projectIdsToUnfreeze)
                 {
                     Contract.ThrowIfFalse(trackerMap.TryGetValue(projectId, out var existingTracker));
-                    // TODO(cyrusn): Is it possible to wrap an underlying tracker with multiple frozen document
-                    // compilation trackers?  Should we be unwrapping as much as we can here?  Or would that also be bad
-                    // given that we're basing what we want to unfreeze on the FrozenSourceGeneratedDocumentStates,
-                    // which may not represent those inner freezes.  Unclear.  There may be bugs here.
-                    var replacingItemTracker = (WithFrozenSourceGeneratedDocumentsCompilationTracker)existingTracker;
-                    trackerMap[projectId] = replacingItemTracker.UnderlyingTracker;
+                    // We unwind all the way, since we could have a chain of frozen source generated documents.
```

Comment:
ensure there are tests for unwrappign multiple times.

and jsut for my understanding, this happens when you do multiple .WithXXX forks on a solution on the generated docs?  like a fork off of a fork?

note: in that case, do you really need to do tracker *wrapping*, or can the second wrap peek into the first and make the changes it needs? 

I ask because unbounded tracker wrapping is something we try to avoid.  it wuold be nice if it were not possible (and WithFrozenSourceGeneratedDocumentsCompilationTracker  could assert it was not passed a WithFrozenSourceGeneratedDocumentsCompilationTracker  as the underlying tracker).

Reply:
Qustion to Jason.  do we need the two trackers anymore? i have found the subclassin there pretty confusing.  would lovei f this functionality actually meant we could go back to a single tracker. perhaps with computed SG docs, or explicitly provided SG docs.  Thoughts?

Reply:
There are tests that cover this, yes, including multiple forks and resetting all the way back.

Jason and I did talk briefly about the long chain of trackers, and my memory was not to worry about it yet. I just took a look though, and its pretty trivial, especially now that I've written the tests and have some (tiny!) understanding of how it works, so I've updated to not create chains.